File: 2020/old/discomplexity/x/js/homeworld2/joint.js
import Matrix from '../x/graphics/matrix.js';

export default class Joint
{
    constructor(name, parent, position, rotation, scale, axis, freedom)
    {
        this.__name = name;
        this.__parent = parent;
        this.__position = position;
        this.__rotation = rotation;
        this.__scale = scale;
        this.__axis = axis;
        this.__freedom = freedom;
        this.__children = new Map();
        this.__objects = new Array();
        this.__model = Matrix.identity();
    }

    get name(){ return this.__name; }
    get parent(){ return this.__parent; }
    get children(){ return this.__children; }
    get objects(){ return this.__objects; }

    render(gl, model, level)
    {
        this.__model = Matrix.translate(model, this.__position);
        if(this.__axis[0] !== 0)
        {
            this.__model = Matrix.rotate(this.__model, this.__axis[0], [1, 0, 0, 0]);
        }
        if(this.__axis[1] !== 0)
        {
            this.__model = Matrix.rotate(this.__model, this.__axis[1], [0, 1, 0, 0]);
        }
        if(this.__axis[2] !== 0)
        {
            this.__model = Matrix.rotate(this.__model, this.__axis[2], [0, 0, 1, 0]);
        }

        if(this.__rotation[1] !== 0)
        {
            this.__model = Matrix.rotate(this.__model, (this.__position[0] > 0 ? this.__rotation[1] : -this.__rotation[1]), [1, 0, 0, 0]);
        }
        if(this.__rotation[0] !== 0)
        {
            this.__model = Matrix.rotate(this.__model, this.__rotation[0], [0, 1, 0, 0]);
        }
        if(this.__rotation[2] !== 0)
        {
            this.__model = Matrix.rotate(this.__model, this.__rotation[2], [0, 0, 1, 0]);
        }

        this.__model = Matrix.scale(this.__model, this.__scale);

        this.__objects.forEach(o => {
            o.render(gl, this.__model, level);
        });
        this.__children.forEach(o => {
            o.render(gl, this.__model, level);
        });
    }
}

File: 2020/old/xgraphics/homeworld2/core/Homeworld2/Graphics/Objects/Joint.cc
#include "Joint.hh"

namespace Homeworld2
{
    namespace Graphics
    {
        namespace Objects
        {
            Joint::Joint(const x::string & name, const x::reality::vector & location, const x::reality::vector & rotation, const x::reality::vector & scale, const x::reality::vector & axis)
            :   x::graphics::objects::joint(location, rotation, scale, axis),
                __name(name)
            {
            }

            Joint::~Joint()
            {
            }

            void Joint::start(const x::reality::matrix & model, const x::reality::vector & center)
            {
                __model = model;

                __model = x::reality::matrix::translate(__model, __location);

                __model = x::reality::matrix::rotate(__model, __axis.x(), {1.0f, 0.0f, 0.0f});
                __model = x::reality::matrix::rotate(__model, __axis.y(), {0.0f, 1.0f, 0.0f});
                __model = x::reality::matrix::rotate(__model, __axis.z(), {0.0f, 0.0f, 1.0f});

                __model = x::reality::matrix::rotate(__model, __location.x() > 0 ? __rotation.y() : -__rotation.y(), {1.0f, 0.0f, 0.0f});__model = x::reality::matrix::rotate(__model, __rotation.x(), {1.0f, 0.0f, 0.0f});
                __model = x::reality::matrix::rotate(__model, __rotation.x(), {0.0f, 1.0f, 0.0f});
                __model = x::reality::matrix::rotate(__model, __rotation.z(), {0.0f, 0.0f, 1.0f});

                __model = x::reality::matrix::scale(__model, __scale);

                __objects.each([this, &center](x::var<x::graphics::object *>::type & o){
                    o->start(__model, center);
                });

                __children.each([this, &center](x::var<x::graphics::objects::joint *>::type & joint){
                    joint->start(__model, center);
                });
            }

            void Joint::update(const x::reality::matrix & model, const x::reality::vector & center)
            {
                __model = model;

                __model = x::reality::matrix::translate(__model, __location);

                __model = x::reality::matrix::rotate(__model, __axis.x(), {1.0f, 0.0f, 0.0f});
                __model = x::reality::matrix::rotate(__model, __axis.y(), {0.0f, 1.0f, 0.0f});
                __model = x::reality::matrix::rotate(__model, __axis.z(), {0.0f, 0.0f, 1.0f});

                __model = x::reality::matrix::rotate(__model, __location.x() > 0 ? __rotation.y() : -__rotation.y(), {1.0f, 0.0f, 0.0f});__model = x::reality::matrix::rotate(__model, __rotation.x(), {1.0f, 0.0f, 0.0f});
                __model = x::reality::matrix::rotate(__model, __rotation.x(), {0.0f, 1.0f, 0.0f});
                __model = x::reality::matrix::rotate(__model, __rotation.z(), {0.0f, 0.0f, 1.0f});

                __model = x::reality::matrix::scale(__model, __scale);

                __objects.each([this, &center](x::var<x::graphics::object *>::type & o){
                    o->update(__model, center);
                });

                __children.each([this, &center](x::var<x::graphics::objects::joint *>::type & joint){
                    joint->update(__model, center);
                });
            }

            void Joint::render(const x::reality::vector & position, const x::reality::vector & center)
            {
                __objects.each([&position, &center](x::var<x::graphics::object *>::type & o){
                    o->render(position, center);
                });

                __children.each([&position, &center](x::var<x::graphics::objects::joint *>::type & joint){
                    joint->render(position, center);
                });
            }
        }
    }
}
